### PR TITLE
Pt 3536

### PR DESCRIPTION
**Pull request by sanufriev**

### CODE REVIEW

File: api/src/main/kotlin/com/rarible/protocol/union/api/service/elastic/EsItemOptimizedSearchService.kt [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+import org.springframework.stereotype.Component
+import java.time.Clock
+import java.time.Instant
+
+@Component
+class EsItemOptimizedSearchService(
+    private val esItemRepository: EsItemRepository,
+    private val esEntitySearchAfterCursorService: EsEntitySearchAfterCursorService,
+    private val properties: EsOptimizationProperties,
+    private val clock: Clock,
```
Comment (yakischik) [inline api/src/main/kotlin/com/rarible/protocol/union/api/service/elastic/EsItemOptimizedSearchService.kt:27]: Why not just nowMillis()?

File: api/src/main/kotlin/com/rarible/protocol/union/api/service/elastic/EsItemOptimizedSearchService.kt [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+        cursor: String?,
+        sort: SortOrder,
+    ): LastUpdatedRange {
+        val from = filterFrom ?: Instant.EPOCH
+        val to = filterTo ?: clock.instant()
+        return when (sort) {
+            ASC -> {
+                val currentFrom = getCursorFrom(cursor) ?: from
+                if (currentFrom == Instant.EPOCH) {
+                    return LastUpdatedRange(from = filterFrom, to = filterTo)
```
Comment (yakischik) [inline api/src/main/kotlin/com/rarible/protocol/union/api/service/elastic/EsItemOptimizedSearchService.kt:96]: Maybe it make sense to put into config date of oldest event in our DB?

File: api/src/main/kotlin/com/rarible/protocol/union/api/service/elastic/EsItemOptimizedSearchService.kt (hunk 1 above)
Comment (sanufriev) [reply api/src/main/kotlin/com/rarible/protocol/union/api/service/elastic/EsItemOptimizedSearchService.kt:27]: Difficult to make a tests

File: api/src/main/kotlin/com/rarible/protocol/union/api/service/elastic/EsItemOptimizedSearchService.kt (hunk 2 above)
Comment (sanufriev) [reply api/src/main/kotlin/com/rarible/protocol/union/api/service/elastic/EsItemOptimizedSearchService.kt:96]: yes, it is a vey good ide

File: api/src/main/kotlin/com/rarible/protocol/union/api/service/elastic/EsItemOptimizedSearchService.kt [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+            sort = sort.sortOrder,
+        )
+        val optimizedFilter = filter.copy(
+            updatedFrom = lastUpdatedRange.from,
+            updatedTo = lastUpdatedRange.to,
+        )
+        val result = esItemRepository.search(optimizedFilter, sort, limit)
+
+        if (result.entities.size >= limit &&
+            result.continuation != null
```
Comment (dmitry-saraykin) [inline api/src/main/kotlin/com/rarible/protocol/union/api/service/elastic/EsItemOptimizedSearchService.kt:77]: || optimizedFilter == filter

**Review by dmitry-saraykin — APPROVED** [review]

**Pull request merged**